source/pele/base_potential.h

```diff
@@ -15,7 +15,32 @@ namespace pele {
 class BasePotential {
 public:
     virtual ~BasePotential() {}
-
+    
+    /**
+     * Return the number of dimensions (box dimensions).
+     * Ideally this should be overloaded.
+     */
+    virtual size_t get_ndim()
+    {
+        throw std::runtime_error("BasePotential::get_ndim must be overloaded");
+    }
+    
+    /**
+     * Return the distance as measured by the distance policy. 
+     */
+    virtual void get_rij(double * const r_ij, double const * const r1, double const * const r2) const
+    {
+        throw std::runtime_error("BasePotential::get_rij must be overloaded");
+    }
+    
+    /**
+     * Return energy_gradient of iteraction. 
+     */
+    virtual double get_interaction_energy_gradient(double r2, double *gij, size_t atom_i, size_t atom_j) const
```

Comment:
ok but somehow we need to pass the distance policy and the interaction
energy_gradient to the pressure_tensor utility.

So the way around this is to add one more hierarchical level and make a
BaseParticlePotential that derives from BasePotential and has this
additional level of complexity. For instance if we wanted an utility that
plots the potential as a function of rij for the particle based potentials,
which would be a good visual test, then we would need that function anyway.
The pressure is in general something that one would want to compute for a
system of particles.

On Thu, May 14, 2015 at 11:16 AM, Jacob Stevenson <notifications@github.com>
wrote:

> In source/pele/base_potential.h
> <https://github.com/pele-python/pele/pull/130#discussion_r30310757>:
>
> > +    {
> > +        throw std::runtime_error("BasePotential::get_ndim must be overloaded");
> > +    }
> > +
> > +    /**
> > +     * Return the distance as measured by the distance policy.
> > +     */
> > +    virtual void get_rij(double * const r_ij, double const * const r1, double const * const r2) const
> > +    {
> > +        throw std::runtime_error("BasePotential::get_rij must be overloaded");
> > +    }
> > +
> > +    /**
> > +     * Return energy_gradient of iteraction.
> > +     */
> > +    virtual double get_interaction_energy_gradient(double r2, double *gij, size_t atom_i, size_t atom_j) const
>
> In general I don't like these changes to BasePotential. What about
> potentials that don't have a concept of distance, or pair interactions,
> e.g. machine learning potentials or lattice based spin potentials. I think
> we should leave BasePotential as simple as possible. you could maybe leave
> get_ndim in, but then you could imagine potentials where the system size
> changes with time e.g. grand canonical.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pele-python/pele/pull/130/files#r30310757>.
>